### PR TITLE
Type NDArrayMixin

### PR DESCRIPTION
**Pull request by Illviljan**
Activate typing on these mixins by removing the Any.

- [ ] Closes #xxxx
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by Illviljan** [comment]
pre-commit.ci autofix

File: xarray/core/utils.py [hunk 1]
```diff
@@ -579,17 +592,17 @@ class NDArrayMixin(NdimSizeLenMixin):
     __slots__ = ()
 
     @property
-    def dtype(self: Any) -> np.dtype:
+    def dtype(self) -> np.dtype:
         return self.array.dtype
 
     @property
-    def shape(self: Any) -> tuple[int]:
```
Comment (Illviljan) [inline xarray/core/utils.py:586]: Anyone understands why `tuple[int]` (single int) was ok before?

**Comment by Illviljan** [comment]
The errors looks like this:

```python
self = <xarray.backends.scipy_.ScipyArrayWrapper object at 0x00000234CBC43610>
variable_name = 'dim2'
datastore = <xarray.backends.scipy_.ScipyDataStore object at 0x00000234CBC43D90>

    def __init__(self, variable_name, datastore):
        self.datastore = datastore
        self.variable_name = variable_name
        array = self.get_variable().data
>       self.shape = array.shape
E       AttributeError: can't set attribute

/home/runner/work/xarray/xarray/xarray/backends/scipy_.py:52: AttributeError
```
Setting shape like this doesn't seem like the proper way of doing it. Maybe these should just inherit NDArrayMixin as well?

**Comment by mathause** [comment]
I think `BackendArray` (and it's parents) did not define `shape` as an attribute. So one option is to revert this change.

---

The other is to unify how the backends assign `shape` and `dtype`? They seem be to defined very differently depending on the backend.

https://github.com/pydata/xarray/blob/d47cf0c850cb70429373782b3c1e0329d14fd05a/xarray/backends/pydap_.py#L28-L30


https://github.com/pydata/xarray/blob/d47cf0c850cb70429373782b3c1e0329d14fd05a/xarray/backends/netCDF4_.py#L59

https://github.com/pydata/xarray/blob/d47cf0c850cb70429373782b3c1e0329d14fd05a/xarray/backends/rasterio_.py#L37

**Comment by Illviljan** [comment]
> I think `BackendArray` (and it's parents) did not define `shape` as an attribute. So one option is to revert this change.

Yeah this approach isn't working that great. I think I'll close this pr unless there's some other ideas.

**Comment by mathause** [comment]
Define 

```python

class BackendArray(...):

     @property 
     def shape(self) -> tuple(int, ...): 
         return self._shape 
```

and then set

```python
class SpecificBackend(BackendArray):

    def __init__(self, variable_name, datastore):
        self._shape = array.shape
```

but that might affect downstream libraries.

**Pull request closed without merging**